crates/sha1/src/sha1.rs

```diff
@@ -0,0 +1,406 @@
+//! Circuits for the [SHA-1] hash function and its internal compression
+//! function.
+//!
+//! [SHA-1]: https://datatracker.ietf.org/doc/html/rfc3174
+
+use bellpepper::gadgets::{multieq::MultiEq, uint32::UInt32};
+use bellpepper_core::{boolean::Boolean, ConstraintSystem, SynthesisError};
+use ff::PrimeField;
+
+use crate::util::{and_uint32, or_uint32};
+
+const ROUND_CONSTANTS: [u32; 4] = [0x5a827999, 0x6ed9eba1, 0x8f1bbcdc, 0xca62c1d6];
+
+const IV: [u32; 5] = [0x67452301, 0xefcdab89, 0x98badcfe, 0x10325476, 0xc3d2e1f0];
+
+pub fn sha1_block_no_padding<Scalar, CS>(
+    mut cs: CS,
+    input: &[Boolean],
+) -> Result<Vec<Boolean>, SynthesisError>
+where
+    Scalar: PrimeField,
+    CS: ConstraintSystem<Scalar>,
+{
+    assert_eq!(input.len(), 512);
+
+    Ok(sha1_compression_function(&mut cs, input, &get_sha1_iv())?
+        .into_iter()
+        .flat_map(|e| e.into_bits_be())
+        .collect())
+}
+
+pub fn sha1<Scalar, CS>(mut cs: CS, input: &[Boolean]) -> Result<Vec<Boolean>, SynthesisError>
+where
+    Scalar: PrimeField,
+    CS: ConstraintSystem<Scalar>,
+{
+    assert!(input.len() % 8 == 0);
+
+    let mut padded = input.to_vec();
+    let plen = padded.len() as u64;
+    // append a single '1' bit
+    padded.push(Boolean::constant(true));
+    // append K '0' bits, where K is the minimum number >= 0 such that L + 1 + K + 64 is a multiple of 512
+    while (padded.len() + 64) % 512 != 0 {
+        padded.push(Boolean::constant(false));
+    }
+    // append L as a 64-bit big-endian integer, making the total post-processed length a multiple of 512 bits
+    for b in (0..64).rev().map(|i| (plen >> i) & 1 == 1) {
+        padded.push(Boolean::constant(b));
+    }
+    assert!(padded.len() % 512 == 0);
+
+    let mut cur = get_sha1_iv();
+    for (i, block) in padded.chunks(512).enumerate() {
+        cur = sha1_compression_function(cs.namespace(|| format!("block {}", i)), block, &cur)?;
+    }
+
+    Ok(cur.into_iter().flat_map(|e| e.into_bits_be()).collect())
+}
+
+fn get_sha1_iv() -> Vec<UInt32> {
+    IV.iter().map(|&v| UInt32::constant(v)).collect()
+}
```

Comment:
The length of the result has a size known at compile time. Have you considered returning an array instead?

Reply:
Thanks for pointing this out. I will try changing the output to an array